File: 0046-permutations/0046-permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        res = []
        chosen = [False for i in range(len(nums))]

        print(chosen)

        def backtrack(curr, chosen):
            if len(curr) == len(nums):
                res.append(curr[:])
                return
            
            for i in range(len(nums)):
                if not chosen[i]:
                    curr.append(nums[i])
                    chosen[i] = True
                    backtrack(curr, chosen)

                    curr.pop()
                    chosen[i] = False
        
        backtrack([], chosen)

        return res
        

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        boolean[] chosen = new boolean[nums.length];
        List<List<Integer>> res = new ArrayList<>();
        dfs(new ArrayList<>(), nums, chosen, res);
        return res;
    }

    private void dfs(List<Integer> perm, int[] nums, boolean[] chosen, List<List<Integer>> res) {
        if (perm.size() == nums.length) {
            res.add(new ArrayList<>(perm));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (!chosen[i]) {
                perm.add(nums[i]);
                chosen[i] = true;
                dfs(perm, nums, chosen, res);
                perm.remove(perm.size() - 1);
                chosen[i] = false;
            }
        }
    }
}